Changed userId value for refresh token events

graziang opened the pull request:
Closes #28567

graziang committed:
Changed userId value for refresh token events

Closes #28567

Signed-off-by: Giuseppe Graziano <g.graziano94@gmail.com>

graziang commented:
@sschu Done. Thanks for the suggestion!

mposolda commented:
@graziang @sschu The change is good, but at the same time, I afraid that in case of `REFRESH_TOKEN_ERROR` because of expired/invalid user session, administrators now don't have any info about the user, who failed to refresh, which previously worked (for non-pairwise subject use-cases).

How about if we possibly have some new details like `refresh_token_sub` and add something like this in TokenManager line 411:
```
if (refreshToken.getSub() != null) {
    event.detail(Details.REFRESH_TOKEN_SUB, refreshToken.getSub());
}
```

What do you think?

sschu commented:
I like that idea @mposolda !

graziang commented:
@mposolda Nice idea also as backwards compatibility. Modified a test to check `refresh_token_sub` and added a small section to release notes. Feel free to edit the section during the #28460

sschu commented on docs/documentation/upgrading/topics/changes/changes-25_0_0.adoc:
```diff
@@ -103,6 +103,13 @@ You can use the `Subject (sub)` mapper to configure the `sub` claim only for acc
+= Changed `userId` for events related to refresh token
+
+The `userId` in the `REFRESH_TOKEN` event is now always taken from user session instead of `sub` claim in the refresh token. The `userId` in the `REFRESH_TOKEN_ERROR` event is now always null.
+The reason for this change is that the value of the `sub` claim in the refresh token may be null with the introduction of the optional `sub` claim or even different from the real user id when using pairwise subject identifiers.
```
You could add "or other ways to override the sub claim" but it is not too important.

graziang replied:
You mean "with the introduction of the optional `sub` claim or even different from the real user id when using pairwise subject identifiers or other ways to override the sub claim"?

sschu replied:
Exactly. In our use case, customers directly take over the sub claim of the external identity provider that is used.

graziang replied:
Sure, makes sense. Done, thanks for the review!

mposolda reviewed (approved):
@sschu @graziang Thanks for the review and updates to this PR!
